pytential/symbolic/primitives.py

```diff
@@ -944,21 +959,63 @@ def _simplex_mapping_max_stretch_factor(ambient_dim, dim=None, dofdesc=None,
             "pd_mat_jtj")
 
     stretch_factors = [
-            cse(sqrt(s), "mapping_singval_%d" % i)
-            for i, s in enumerate(
-                _small_sym_mat_eigenvalues(
-                    # Multiply by 4 to compensate for equilateral reference
-                    # elements of side length 2. (J^T J contains two factors of
-                    # two.)
-                    4 * equi_pder_mat_jtj))]
+            cse(sqrt(s), f"simplex_mapping_singval_{i}")
+            # NOTE: multiply by 4 to compensate for equilateral reference
+            # elements of side length 2. (J^T J contains two factors of two.)
+            for i, s in enumerate(_small_sym_mat_eigenvalues(4 * equi_pder_mat_jtj))
+            ]
+
+    from pymbolic.primitives import Max
+    result = Max(tuple(stretch_factors))
+
+    if with_elementwise_max:
+        result = ElementwiseMax(result, dofdesc=dofdesc)
+
+    return result
+
+
+def _hypercube_mapping_max_stretch_factor(ambient_dim, dim=None, dofdesc=None,
+        with_elementwise_max=True):
+    if dim is None:
+        dim = ambient_dim - 1
+
+    # NOTE: unlike in the simplex case, here we do not need to transform the
+    # reference element, as it already is nicely rotation invariant
+    pder_mat = first_fundamental_form(ambient_dim, dim=dim, dofdesc=dofdesc)
+    stretch_factors = [
+            cse(sqrt(s), f"hypercube_mapping_singval_{i}")
+            # NOTE: like for the simplex, we multiply `pder_mat` by 4 to
+            # account for the side length of 2 in J^T J
+            for i, s in enumerate(_small_sym_mat_eigenvalues(4.0 * pder_mat))
+            ]
 
     from pymbolic.primitives import Max
     result = Max(tuple(stretch_factors))
 
     if with_elementwise_max:
         result = ElementwiseMax(result, dofdesc=dofdesc)
 
-    return cse(result, "mapping_max_stretch", cse_scope.DISCRETIZATION)
+    return result
+
+
+def _mapping_max_stretch_factor(ambient_dim, dim=None, dofdesc=None,
+                                with_elementwise_max=True):
+    simplex_stretch_factor = _simplex_mapping_max_stretch_factor(
+        ambient_dim, dim, dofdesc=dofdesc,
+        with_elementwise_max=with_elementwise_max)
+    hypercube_stretch_factor = _hypercube_mapping_max_stretch_factor(
+        ambient_dim, dim, dofdesc=dofdesc,
+        with_elementwise_max=with_elementwise_max)
+
+    import modepy as mp
+    from pymbolic.primitives import If
+    stretch_factor = If(IsShapeClass(mp.Simplex, dofdesc),
+                        simplex_stretch_factor,
+                        If(IsShapeClass(mp.Hypercube, dofdesc),
+                           hypercube_stretch_factor, 0)
```

Comment:
Is there a nice way to go `else raise` in here? I tried adding a `var("__unreacheable")`, but the compiler was complaining about it :(

Reply:
`Quotient(1, 0)`? :grin:

Reply:
Maybe make an `Error(string)` symbol?